velox/functions/lib/aggregates/MinMaxAggregateBase.h

```diff
@@ -0,0 +1,646 @@
+      bool mayPushdown) override {
+    addSingleGroupRawInput(group, rows, args, mayPushdown);
+  }
+};
+
+template <
+    template <typename T>
+    class TNumeric,
+    template <bool compareStopAtNull>
+    typename TNonNumeric,
+    typename TTimestampFunc>
+exec::AggregateFunctionFactory getMinMaxFunctionFactoryInternal(
+    const std::string& name,
+    bool nestedNullAllowed) {
+  auto factory = [name, nestedNullAllowed](
+                     core::AggregationNode::Step step,
+                     std::vector<TypePtr> argTypes,
+                     const TypePtr& resultType,
+                     const core::QueryConfig& /*config*/)
+      -> std::unique_ptr<exec::Aggregate> {
+    const bool throwOnNestedNulls =
+        !nestedNullAllowed && velox::exec::isRawInput(step);
+    auto inputType = argTypes[0];
+    switch (inputType->kind()) {
+      case TypeKind::BOOLEAN:
+        return std::make_unique<TNumeric<bool>>(resultType);
+      case TypeKind::TINYINT:
+        return std::make_unique<TNumeric<int8_t>>(resultType);
+      case TypeKind::SMALLINT:
+        return std::make_unique<TNumeric<int16_t>>(resultType);
+      case TypeKind::INTEGER:
+        return std::make_unique<TNumeric<int32_t>>(resultType);
+      case TypeKind::BIGINT:
+        return std::make_unique<TNumeric<int64_t>>(resultType);
+      case TypeKind::REAL:
+        return std::make_unique<TNumeric<float>>(resultType);
+      case TypeKind::DOUBLE:
+        return std::make_unique<TNumeric<double>>(resultType);
+      case TypeKind::TIMESTAMP:
+        return std::make_unique<TTimestampFunc>(resultType);
+      case TypeKind::HUGEINT:
+        return std::make_unique<TNumeric<int128_t>>(resultType);
+      case TypeKind::VARBINARY:
+        [[fallthrough]];
+      case TypeKind::VARCHAR:
+        return std::make_unique<TNonNumeric<false>>(inputType, false);
+      case TypeKind::ARRAY:
+        [[fallthrough]];
+      case TypeKind::ROW:
+        if (nestedNullAllowed) {
+          return std::make_unique<TNonNumeric<false>>(
+              inputType, throwOnNestedNulls);
+        } else {
+          return std::make_unique<TNonNumeric<true>>(
+              inputType, throwOnNestedNulls);
+        }
+      case TypeKind::UNKNOWN:
+        return std::make_unique<TNumeric<UnknownValue>>(resultType);
+      default:
+        VELOX_UNREACHABLE(
+            "Unknown input type for {} aggregation {}",
+            name,
+            inputType->kindName());
+    }
+  };
+  return factory;
+}
+
+/// Min & Max functions in Presto and Spark have different semantics:
+/// 1. Nested nulls are allowed in Spark but not Presto.
+/// 2. The timestamp type represents a time instant in microsecond precision in
+/// Spark, but millis precision in Presto.
+/// We add parameter 'nestedNullAllowed'
+/// and template TTimestampAggregate to register min and max functions with
+/// different behaviors.
+template <typename TTimestampMinAggregate = MinAggregate<Timestamp>>
```

Comment:
It seems that the only different is timestamp precision. Let's pass that as an argument (TimestampPrecision enum).

Then, we can move the rest of this file into .cpp.

Reply:
Updated, thanks